src/OrbitGl/GraphTrack.cpp

```diff
@@ -155,61 +187,109 @@ void GraphTrack::DrawLabel(Batcher& batcher, TextRenderer& text_renderer, Vec2 t
-                 layout_->GetSpaceBetweenTracksAndThread() + layout_->GetEventTrackHeight() +
-                 layout_->GetTrackBottomMargin();
-  return height;
 }
 
-void GraphTrack::SetLabelUnitWhenEmpty(const std::string& label_unit) {
-  if (!label_unit_.empty()) return;
-  label_unit_ = label_unit;
-}
-
-void GraphTrack::SetValueDecimalDigitsWhenEmpty(uint8_t value_decimal_digits) {
-  if (value_decimal_digits_.has_value()) return;
-  value_decimal_digits_ = value_decimal_digits;
-}
+template <size_t Dimension>
+void GraphTrack<Dimension>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                       float z) {
+  auto entries_within_time_range = series_.GetEntriesInTimeRange(min_tick, max_tick);
+  if (!entries_within_time_range.has_value()) return;
+
+  // Draw rectanges according to the time series entries located in [min_tick, max_tick].
+  double min = series_.min_series_value();
+  double inverse_value_range = series_.inverse_value_range();
+  auto it = entries_within_time_range.value().begin;
+  uint64_t previous_time = it->first;
+  std::array<float, Dimension> previous_normalized_values;
+  std::transform(it->second.begin(), it->second.end(), previous_normalized_values.begin(),
+                 [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  for (++it; it != entries_within_time_range.value().end; ++it) {
+    uint64_t current_time = it->first;
+    DrawSingleSeriesEntry(batcher, previous_time, current_time, previous_normalized_values, z);
+
+    previous_time = current_time;
+    std::transform(it->second.begin(), it->second.end(), previous_normalized_values.begin(),
+                   [min, inverse_value_range](double value) {
+                     return static_cast<float>((value - min) * inverse_value_range);
+                   });
+  }
 
-void GraphTrack::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
-  constexpr uint32_t kDepth = 0;
-  std::shared_ptr<TimerChain> timer_chain = timers_[kDepth];
-  if (timer_chain == nullptr) {
-    timer_chain = std::make_shared<TimerChain>();
-    timers_[kDepth] = timer_chain;
+  uint64_t time_of_first_entry_within_time_range = entries_within_time_range.value().begin->first;
+  if (min_tick > series_.StartTimeInNs() && min_tick < time_of_first_entry_within_time_range) {
+    const std::array<double, Dimension>& values = series_.GetValuesOfPreviousOrFirstEntry(min_tick);
+    std::transform(values.begin(), values.end(), previous_normalized_values.begin(),
+                   [min, inverse_value_range](double value) {
+                     return static_cast<float>((value - min) * inverse_value_range);
+                   });
+    DrawSingleSeriesEntry(batcher, min_tick, time_of_first_entry_within_time_range,
+                          previous_normalized_values, z);
   }
```

Comment:
All of this made the code more complicated than easier. The idea of `GetEntriesInTimeRange` was an attempt to make this method easier. I actually prefer the first version over this. 

I implemented something you could be inspired by: https://paste.googleplex.com/6502721508081664

Reply:
Thanks Anton! I like your suggestion. But I'm not sure about the name `GetEntriesInTimeRange`.

For example if the time series contains three entries with timestamp 0, 10, and 20. And our time range is [5, 15]. From the method name it seems to return only the entry with timestamp 10, but actually it returns all three entires.

Reply:
Yes, I agree, the name `GetEntriesInTimeRange` is then not good anymore. Please take a look in the gpaste. I also put another name suggestion there: `GetEntriesAffectedByTimeRange`. I dont know if this is a perfect name, so a comment explaining what the function does exactly, is probably a good idea. 

You can also leave the code as currently is, but I think my suggestion makes a lot of sense. In my opinion the DrawSeries function gets easier to understand and its shrinking the size of the function from 47 to 24.

Reply:
Thanks for the suggestion and I changed the current code accordingly. 

I made a small modification on `GetNextOrEndIter` as it returns the iterator to the next next entry. I changed it as `GetNextOrLastEntry` which returns iterator to the next entry if exists, otherwise returns iterator to the last entry.